rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1059 @@
+  - The cutoff is usually determined empirically based on common usage patterns.
+
+**Examples:**
+
+```nix
+{
+  #1
+  buildInputs = [
+    foo
+    bar
+    baz
+
+    somethingElse
+  ];
+
+  #2
+  systemd.services = {
+    foo = { };
+    bar = { };
+  };
+
+  #3
+  inherit
+    lib
+    foo
+    bar
+    baz
+  ;
+}
+```
+
+### Function application
+
+**Description:**
+
+- In a function application chain, the first element is treated as the "function" and the remaining ones as "arguments".
+- The last argument receives special treatment, to better represent common coding patterns.
+- As much arguments as possible are fit onto the first line.
+  - If all but the last argument do fit, then the last argument may start on the same line.
+  - If an earlier argument does not fit onto the first line, then itself and all the following ones start on a new line. This is called the expanded form.
+  - All arguments that are not on the same line as the function are indented by one level.
+
+**Examples:**
+
+```nix
+#1
+function arg1 arg2
+
+#2
+function arg1 arg2 {
+  more = "things";
+}
+
+#3
+function arg1 arg2 arg3 # reached line limit here
+  arg4
+  arg5
+  [
+    1
+    2
+  ]
+  arg7
+
+#4 All arguments that don't fit on the same line, start indented on a new line
+function arg1
+  {
+    # items that don't fit on same line
+  }
+  arg3
+
+#5
+function
+  {
+    a = 1;
+    b = 2;
+  }
+  {
+    c = 1;
+    d = 2;
+  }
+
+#6 (with "--width=20")
+function arg1 (
+  function2 args
+)
+```
+
+**Drawbacks**
+
+- This style sometimes forces lists or attribute sets to start on a new line, with additional indentation of their items.
+
+**Rationale and alternatives**
+
+- Not indenting the arguments, to save some indentation depth. This would be consistent with other constructs like function declarations and let bindings.
+- Compacting multiline arguments like this:
+  ```nix
+  #4b
+  function arg1 {
+    # stuff
+  } arg3
+
+  #5b
+  function {
+    # …
+  } {
+    # …
+  }
+  ```
+  - This violates the guideline of the indentation representing the expression structure, and thus reduces readability.
+  - This does not work well with line length limits on short arguments like in example #3.
+
+### Function declaration
+
+**Description**
+
+- The body of the function is not indented relative to its arguments.
+- Multiple ("simple") identifier arguments are written onto the same line if possible.
+- Attribute set arguments always start on a new line; they are not mixed with identifier arguments.
+  - If they have few attributes, the argument may be written on a single line, otherwise the expanded form is used.
+- Attribute set arguments have their attributes on a new line each with indentation, followed by a trailing comma.
+
+**Examples**
+
+```nix
+#1
+name: value: name ++ value
+
+#2
+name: value:
+{
+  "${name}-foo" = value;
+}
+
+#3
+{ pkgs }: pkgs.hello
+
+#4
+args@{
+  some,
+  argument,
+  default ? value,
+  ...
+}:
```

Comment:
_I know this is a holy war, but my eyes and brain refuse to read such arguments without leading commas... And for a reason (described below)._

__Leading aligned commas serve additional information for quicker recognition by a human.__
You see a geometric structure quicker, where comma position is fixed at the "x" axis. This gives quicker information about whether each line specifies an input attribute and whether its syntax is correct.

```nix
{ some
, argument
, some-other
, default ? value
, ...
}@args:
```

With leading commas the algorithm to parse a single line of an attrset-argument syntax with eyes is as follows:
1. locate a vertical line of strokes (`,` `{` `}`) on the left
2. isolate a continuous set of characters apart from whitespace
3. parse an attribute name

![nix-rfc0166](https://github.com/NixOS/rfcs/assets/464913/dc8e1702-91b7-4e1f-b497-f61d27f4e2b8) ![nix-rfc0166-myscr_2023-11-17T03:51:48+04:00](https://github.com/NixOS/rfcs/assets/464913/888df73e-e408-40e3-af2e-e340e3a81eab)

Human brains are fast at tracing lines and highlighting words on a solid background. Just a single operation is needed to validate most of the attrset-argument syntax.

---

```nix
args@{
  some,
  argument,
  some-other,
  default ? value,
  ...,
}:
```

With trailing commas the algorithm to parse a single line of an attrset-argument syntax with eyes is as follows:
1. go to a line above
2. traverse it to the end
3. check whether the last character is either:
    - `{`
    - `,`
4. go back to a line below
5. traverse the current line to the end
6. check whether the last character is `,`
7. cut a continuous set of characters, which:
    - begins after ` `
    - ends before `,` (or also ` ` if trailing commas are not forced)
8. parse an attribute name

This not only applies to parsing, but navigating to the end of each line to write a single character if it's not in place.

Now lets see how people with poor vision can misinterpret the trailing comma pattern in Nix language:
![image](https://github.com/NixOS/rfcs/assets/464913/9a4ad4f9-c0f9-497b-8e72-8a991b4aedc9)
Is it a list like this?
```nix
[
  arg-checksum
  argument-cpufreq
  program
  chromium
]
```

But with leading commas you would never confuse it with a list:
![image](https://github.com/NixOS/rfcs/assets/464913/3b703044-1ed4-4de4-864a-a5354a8801a9)

---

Some benefits of leading commas:
1. No need to traverse each line to the end.
3. No need to check for scattered unaligned commas.
4. Simpler rule to discern attribute name (simply surrounded by whitespace).
5. Better distinction of attrset-arguments with other elements of the language (e.g. never confusing attrset-arguments with attrset definition, which uses visually similar `;` character).
6. Not breaking nixpkgs code base and originality, where almost each invocation of something like `nix edit nixpkgs#hello` shows the well-distinguished geometric pattern of package dependencies.

Yes, there are valid points for leading commas, mainly being diff-friendly. But everything cannot be ideal. Sacrificing so much of code readability for a single line diff benefit does not sound optimal, IMHO. The mentioned problems of editors, doing wrong folding of comments with leading commas, is a tooling problem and not a language formatting.
If the first line diff is considered so important, maybe add support for leading commas (similarly to trailing commas) like this:
```nix
{
, some
, argument
, some-other
, default ? value
, ...
}@args:
```

Reply:
My personal requirements for being okay with leading commas are:

1. Propose a good format that works with per argument comments, multiline comments, multiline arguments. Simply liking leading comments in the trivial cases is not sufficient.
2. Have support for leading comma on the first argument in the language

Another argument, albeit a more idealistic one, is that a format should be independent of the indentation size: leading commas would look weird for any indentation other than two space

Reply:
If we take a look around,
barely any other language uses leading commas. Typing leading commas is thus unintuitive for most people coming from other languages. 

Haskellers have developed the idiom of leading comma.

Reply:
To add to that, all languages I know of in which leading comma is used, either don't or didn't support trailing commas on the last item. Leading commas are always a workaround as the less-bad option when there are no trailing commas.

Reply:
We looked at this in today's shepherd meeting. We are convinced that the trailing commas are overall better, which also seems to align with most people's reactions in this thread, so we intend to keep the RFC as is.

We did however expand the argumentation in the RFC to also include the good new arguments given in this thread: https://github.com/nix-rfc-101/rfcs/commit/4f3b3ce5d704755d6527a9eaae9d01b9b430104d

Reply:
### trying to structure comments in this thread

So, the comments above with criticism of the current leading style formatting can be structured into these theses:

1. leading commas style doesn't have a complete proposal with enough cases documented
2. leading commas are _unintuitive_ or unusual comparing to other languages
3. leading commas are a historical workaround _("less-bad option when there are no trailing commas")_
4. trailing commas seem to _"align with most people's reactions in this thread"_

I think thesis [1] is objective and valid. And a complete proposal has to be defined. I try to address some points below. Yes, it's not always trivial, with compromises in some cases. But there is no ideal solution with trailing commas either.

As for thesis [2] and [3] it's hard to argue, because being quite subjective.
Except the comparison with other languages, where we can admit statistical difference, but not the outcome it brings. Is being similar always a good point or being different always a bad point?

As for the thesis [4] the situation is not as clear as stated (@infinisil):
![image](https://github.com/NixOS/rfcs/assets/464913/3ee3cec5-36ae-4191-aa24-fe22edc40afa)
Even here we have 50/50%. Many contributors do not participate in reading RFC 0166, do not vote. And I think such a drastic change in nixpkgs formatting may come as a shock to most.

---

## a few more arguments in favor of leading commas

I am very sorry for not providing exact subjective statements this time below, but, IMHO, these are worth your attention too :pray:.

### unusual solutions to unusual challenges

> barely any other language uses leading commas

I think our answer to this:
_barely any other language solves problems which Nix is designed to solve_.

I believe it's **not a coincidence** to have decisions which look weird compared to other tools.

Other conventional languages are not used to **de-sugar input attribute sets** (say structures/records) to this extent as Nixpkgs does in practice. Leading commas is one of the most used syntax patterns, since we deal with **lots of input entities** in a single function most of the time. I cannot believe we can trade the readability of the cornerstone input Attribute Set language construct for one line diff, unadapted editors, etc.

Nix ecosystem is on the rise. Lets stabilize things which are not blockers and concentrate on those which are (e.g. flakes). Leading commas did not stop us from creating the biggest software repository on the Earth.

### Nixpkgs historical case

Forcing Nixpkgs to get rid of leading commas is the biggest break to the part of its established syntax style. I would say it's the Nixpkgs mark, which you see in each package definition. Out of all syntax patterns in the RFC the leading commas pattern is the most used and the most stable one in Nixpkgs (and not only there). The debate is kinda similar to the established `nix shell` naming, which has polished through years (losing which would mean a huge loss for our community (https://github.com/NixOS/nix/issues/4715#issuecomment-1002161350)).

### first argument diff in nixpkgs

Usually, package is dependent on stdenv or lib, so these are placed first.
If we take this as a rule, most first argument diffs will be neutralized. Surely, this is a compromise as it covers top-level package functions mostly. But almost any part of the decision to take on formatting will be a compromise regardless of the comma position.

### tooling

Still, implementing a language support for leading comma on the first argument is most likely an overkill.
**Tooling is the key** in my opinion. Lets add support for established format to editors, rather than trying to adapt code of one of the largest and most active repositories in the world to them at this enormous cost.

### let's briefly go through the points of criticism of leading commas style in RFC

* _"Moving items around with this style may require editing lines."_
    True with the first item. Replacing letter `{` with `,` and vice versa. 
    An editor plugin could automate this if it's a problem. Such diffs will be rare in nixpkgs.
* _"Inconsistent with the [expression expansion guidelines](https://github.com/NixOS/rfcs/blob/463a858e57e4d68f4bfb1d38f20d8ac852ac4574/rfcs/0166-nix-formatting.md#expansion-of-expressions), which disallows forms like { some; some should start on a new line instead."_
    As I see, the core idea of the guidelines regarding "_each item must be on its own line_" is the ease of items begin search and vertical space consistency. Leading commas are consistent in having special character (`{` or `,`) in the beginning of each line.
    The current wording, however, does not strictly prevent other symbols to appear on the same line. Except the counter-example in the subsection of the rule: _"e.g. [ firstElement in a multi line list is not allowed"_.
    So, just a slightly different wording for this rule will be:
    * each line must contain at most one item
    * all lines must begin vertically aligned with the same type of character (e.g. letter, special)
    * _(additionally)_ all items must begin vertically aligned between each other
* _"This does not work well with leading @ bindings."_
    Use trailing @ bindings. What are the downsides?
* _"It's unclear whether comments belong to the next or the previous argument."_
    Rule: comment (if any) must start before argument's leading comma.
* _"The leading comma style was a lesser-evil workaround for the lack of trailing commas in the Nix language..."_
    In my opinion, this point is only a historical one, without any influence on our current decision process. Not all available features in the language need to be used.

### testbed

> For any proposal, please use https://github.com/NixOS/nixpkgs/blob/a1ed34a3161eb27eab6c743a54ce0dbbe6db0b21/pkgs/stdenv/generic/make-derivation.nix#L105-L203 as a benchmark. It covers almost all relevant corner cases

Nice one! It is quite a good example of how leading commas help to orient among many attribute set arguments. Personally, I like most of it and just need to formalize and standardize the decisions made there to address:
> format that works with per argument comments, multiline comments, multiline arguments

@piegamesde, btw, is there already a sample of this benchmark, formatted in the proposed trailing commas style?

---

_I hope I did not offend anybody :pray:. I highly appreciate the work of all the contributors and I'm very grateful to Nix community in general for doing such a great piece of tech and giving me an opportunity to have a Nix job._

Reply:
To me, your argument in favor of leading commas boils down to "we can do this". Yes, absolutely. All of these issues are solvable and might have a decent solution. We are making the rules right now, we absolutely can change these rules if we can think of better ones.

But if you take a step back, and ignore all historic path dependencies and any personal "I just prefer it like this" preferences, and have a look at both options and neutrally compare them—do you still think that having leading commas in function declarations is preferable over trailing commas? At least to me, the choice is pretty easy: One of them comes with lots of special cases and caveats and potential inconsistencies and special requirements for tooling support, the other one is relatively straightforward.

-------

You can have a look at the current Nixmft output here: https://github.com/piegamesde/nixpkgs/blob/nixfmt-2/pkgs/stdenv/generic/make-derivation.nix#L76-L182. Keep in mind that the formatter is constrained by the input in a way a human adapting the code would not be, and also that any vertical alignment got removed in the process, which is a different topic. After some minor manual cleanup, it could look like this (note that I chose to reintroduce the vertical alignment, to make it closer to the original, so that we may better ignore that aspect in the comparison):

```nix
  mkDerivationSimple =
    overrideAttrs:

    # `mkDerivation` wraps the builtin `derivation` function to
    # produce derivations that use this stdenv and its shell.
    #
    # See also:
    #
    # * https://nixos.org/nixpkgs/manual/#sec-using-stdenv
    #   Details on how to use this mkDerivation function
    #
    # * https://nixos.org/manual/nix/stable/expressions/derivations.html#derivations
    #   Explanation about derivations in general
    {
      # These types of dependencies are all exhaustively documented in
      # the "Specifying Dependencies" section of the "Standard
      # Environment" chapter of the Nixpkgs manual.

      # TODO(@Ericson2314): Stop using legacy dep attribute names

      #                           host offset -> target offset
      depsBuildBuild ? [ ], #              -1 -> -1
      depsBuildBuildPropagated ? [ ], #    -1 -> -1
      nativeBuildInputs ? [ ], #           -1 ->  0  N.B. Legacy name
      propagatedNativeBuildInputs ? [ ], # -1 ->  0  N.B. Legacy name
      depsBuildTarget ? [ ], #             -1 ->  1
      depsBuildTargetPropagated ? [ ], #   -1 ->  1

      depsHostHost ? [ ], #                 0 ->  0
      depsHostHostPropagated ? [ ], #       0 ->  0
      buildInputs ? [ ], #                  0 ->  1  N.B. Legacy name
      propagatedBuildInputs ? [ ], #        0 ->  1  N.B. Legacy name

      depsTargetTarget ? [ ], #             1 ->  1
      depsTargetTargetPropagated ? [ ], #   1 ->  1

      checkInputs ? [ ],
      installCheckInputs ? [ ],
      nativeCheckInputs ? [ ],
      nativeInstallCheckInputs ? [ ],

      # Configure Phase
      configureFlags ? [ ],
      cmakeFlags ? [ ],
      mesonFlags ? [ ],
      # Target is not included by default because most programs don't care.
      # Including it then would cause needless mass rebuilds.
      #
      # TODO(@Ericson2314): Make [ "build" "host" ] always the default / resolve #87909
      configurePlatforms ?
        lib.optionals (stdenv.hostPlatform != stdenv.buildPlatform || config.configurePlatformsByDefault)
          [
            "build"
            "host"
          ],

      # TODO(@Ericson2314): Make unconditional / resolve #33599
      # Check phase
      doCheck ? config.doCheckByDefault or false,

      # TODO(@Ericson2314): Make unconditional / resolve #33599
      # InstallCheck phase
      doInstallCheck ? config.doCheckByDefault or false,

      # TODO(@Ericson2314): Make always true and remove / resolve #178468
      strictDeps ?
        if config.strictDepsByDefault then true else stdenv.hostPlatform != stdenv.buildPlatform,

      enableParallelBuilding ? config.enableParallelBuildingByDefault,

      meta ? { },
      passthru ? { },
      # position used in error messages and for meta.position
      pos ? (
        if attrs.meta.description or null != null then
          builtins.unsafeGetAttrPos "description" attrs.meta
        else if attrs.version or null != null then
          builtins.unsafeGetAttrPos "version" attrs
        else
          builtins.unsafeGetAttrPos "name" attrs
      ),
      separateDebugInfo ? false,
      outputs ? [ "out" ],
      __darwinAllowLocalNetworking ? false,
      __impureHostDeps ? [ ],
      __propagatedImpureHostDeps ? [ ],
      sandboxProfile ? "",
      propagatedSandboxProfile ? "",

      hardeningEnable ? [ ],
      hardeningDisable ? [ ],

      patches ? [ ],

      __contentAddressed ?
        (!attrs ? outputHash) # Fixed-output drvs can't be content addressed too
        && config.contentAddressedByDefault,

      # Experimental.  For simple packages mostly just works,
      # but for anything complex, be prepared to debug if enabling.
      __structuredAttrs ? config.structuredAttrsByDefault or false,

      env ? { },

      ...
    }@attrs:
```

Reply:
why do this
```nix
{ foo
, bar
, foobar
}:
```
but not this?
```nix
{ foo = "foo"
; bar = "bar"
; foobar = "foobar"
;}
```

Reply:
@quantenzitrone indeed, both syntax constructs are related. What you point out is precisely the converse of our current rules: *Because* leading `;` look so bad, we also forbid leading `,` for the sake of consistency.

By the way, I am not opposed to leading item separators instead of trailing ones. However, my requirement for this would be to allow a leading `,` on the very first item. (Then, this style would be equivalent to simply having a bullet point list except that it starts with `,` instead of `-`)

Reply:
> - on the contrary, with leading commas the solution is simple:
>   - align `,` on the left
>   - align `#` on the right

@AleXoundOS I'm not sure what exactly you mean by this, could you please expand and maybe also provide some examples?

Reply:
@piegamesde, I mean this:

### trailing commas unaligned

```
  nativeBuildInputs ? [ ], #           -1 ->  0  N.B. Legacy name
  propagatedNativeBuildInputs ? [ ], # -1 ->  0  N.B. Legacy name
  depsBuildTarget ? [ ], #             -1 ->  1
  depsBuildTargetPropagated ? [ ], #   -1 ->  1
```

### trailing commas with `#` aligned

```
  nativeBuildInputs,           # -1 ->  0  N.B. Legacy name
  propagatedNativeBuildInputs, # -1 ->  0  N.B. Legacy name
  depsBuildTarget,             # -1 ->  1
  depsBuildTargetPropagated,   # -1 ->  1
```

### leading commas with both `,` and `#` aligned

```
, nativeBuildInputs           # -1 ->  0  N.B. Legacy name
, propagatedNativeBuildInputs # -1 ->  0  N.B. Legacy name
, depsBuildTarget             # -1 ->  1
, depsBuildTargetPropagated   # -1 ->  1
```

### trailing commas with both `,` and `#` aligned (weird, IMHO)

```
  nativeBuildInputs          , # -1 ->  0  N.B. Legacy name
  propagatedNativeBuildInputs, # -1 ->  0  N.B. Legacy name
  depsBuildTarget            , # -1 ->  1
  depsBuildTargetPropagated  , # -1 ->  1
```

Reply:
2, 3 and 4 would not work with the vertical alignment nonpreservation rule

https://github.com/nix-rfc-101/rfcs/blob/463a858e57e4d68f4bfb1d38f20d8ac852ac4574/rfcs/0166-nix-formatting.md?plain=1#L437

It may be good to specifically point out that vertical alignment of comments is not preserved as well.

Either there or in the section about comments: https://github.com/nix-rfc-101/rfcs/blob/463a858e57e4d68f4bfb1d38f20d8ac852ac4574/rfcs/0166-nix-formatting.md?plain=1#L590C1-L607

Reply:
We've discussed this again in the shepherd team but still stand by our decision to use trailing commas. Thanks for your input regardless though!

Reply:
@infinisil, thank you for paying attention to this! Let's see how it goes. But, even considering the current spontaneous votes status ![image](https://github.com/NixOS/rfcs/assets/464913/9e13a1cd-75c0-4929-a626-698f740e679e) (under the first message of this thread), 54% does not seem to be enough for a radical change to happen and community debate seems to continue as change begins.

Reply:
This is not how our decision making process works.

Reply:
*TL;DR: Everything I've already said before and also written into the RFC, but in more detail and with many more words.*

## Leading commas

Okay, so let's design a format for function arguments and leading commas and try to make it as good as possible. I'll cover all relevant design questions. (More may come up during an actual implementation attempt, but this should cover most of the ground.)

### 1. Basics

Let's start simple: Just three arguments, all single-line. There are already two possibilities to choose from.

```nix
# 1
{
  arg1
, arg2 ? null
, arg3
}:

# 2
{ arg1
, arg2 ? null
, arg3
}:

[]
```

You may say "obviously we want the second one", but this one contradicts our current expansion rules:

> - If expanded into multiple lines, each item must be on its own line.
>   - This also applies to the first item, so e.g. `[ firstElement` in a multi line list is not allowed.

I agree though that the second style is preferable, so now another choice has to be made: Adapt the other rules derived from or accept the inconsistency.

```nix
[ first
  second
  third
]

{ one = "1";
  two = "2";
  tree = "3;
}
```

Taking that style further quickly leads to chaos and further inconsistencies:

```nix
rec {
  one = "1";
  two = "2";
  three = "3";
}

function call [
  first
  second
  third
]

function call # long line
  [ first
    second
    third
  ]
```

This may even have ripple effects onto other syntax constructs like parentheses. So it looks like making only one exception to the expression expansion rule is the less bad option here.

### 2. Multiline arguments

Things are already getting tricky.

```nix
# 1
{ one ? [
    1
    1
  ]
, two ?
    if "two" == 2 then 2 else "two"
, three ? function [
    1
    2
  ]
}:

# 2
{ one ? [
  1
  1
]
, two ?
    if "two" == 2 then 2 else "two"
, three = function [
  1
  2
]
}:

# 3
{ one ?
  [
    1
    1
  ]
, two ?
  if "two" == 2 then 2 else "two"
, three ?
  function [
    1
    2
  ]
}:

# 4
{ one
  ? [
    1
    1
  ]
, two
  ? if "two" == 2 then 2 else "two"
, three
  ? function [
    1
    2
  ]
}:

[]
```

There are many other variations to handle the indentation here, but it quickly becomes clear that the base indent must be at +2 of the start of the function argument name, or –equivalently– at +4 of the start of its line. This is in violation of our "double indentation" rule, which states that from one line to the next one may only indent upt to a single level. Currently, that rule is the only "hard" rule in the RFC style, with absolutely zero exceptions.

The only one which avoids this to be somewhat taken seriously is option \#4, but it is as desperate an attempt as it looks. It has several downsides, from diff noise when switching between multi-line and single-line defaults to weird interactions with function calls (not shown in this example).

So I'll accept defeat and make one exception to the sacred double indentation rule, and proceed with style \#1.

### 3. `@`-pattern

In Nix, you can have the `@`-pattern either before or after the argument destructuring. Both are equivalent.

```nix
# 1
args@{
  one
, two
, three
}:

# 2
{ one
, two
, three
}@args:

# 3
{ one
, two
, three
}
@args:

# 4
{ one
, two
, three
}
@ args:

# 5
{ one
, two
, three
} @ args:

[]
```

Currently in Nix leading `@` is the more prevalent convention (estimated through some quick greps through `nixpkgs`), but clearly it doesn't work here: the first argument without a leading comma just looks too inconsistent.

So the only sensible option is to normalize them to the trailing version during formatting. (Note that this is not as trivial as it may look like at a first glance, since any token may have comments associated with it which need proper handling as well. But it should be doable.)

There are lots of way about how to handle the whitespace for trailing `@`-patterns, but on a cursory glance this looks like a mostly subjective decision so I'll arbitrarily pick style \#4.

### 4. Comments

Arguments may have comments associated with them. These may come in various forms, `#`, `/*`, `/**`, single-line or spanning multiple lines. Trailing comments on a line are trivial and don't need to be discussed.

```nix
# 1
{ /* comment */
  one
, /*
    multiline
    comment
  */
  two
, # comment
  three
, # multiline
  # comment
  four
}:

# 2
/* comment */
{ one
/*
  multiline
  comment
*/
, two
/* comment */
, three
# multiline
# comment
, four
}:

[]
```

Style \#1 has the disadvantage that the comma gets increasingly distanced from the argument it is associated with (comments may be long). It also loses the clear vertical line to the left which is one of the big advantages of a leading comma style.

Style \#2 looks more promising though. When using `#` comments it even has a clear vertical line for visual reference. It may be tempting to thus decide to normalize all `/*` comments to `#` for that purpose. But this may be a controversial change on its own, but more importantly this cannot be done for `/**` doc comments according to RFC 145. The RFC does not specify any `#`-starting doc comments, so this idea is out. Additionally, style \#2 has the issue that the first comment collides with a potential comment for the entire function. This is not compatible with RFC 145 either.

Therefore, using style \#1 for comments is the only viable option here.

## Trailing commas

Okay and now let's do the same thing again, but with trailing commas.

### 1. Basics

Same decision as last time:

```nix
# 1
{
  arg1,
  arg2 ? null,
  arg3,
}:

# 2
{ arg1,
  arg2 ? null,
  arg3,
}:

[]
```

This time it is easy though, we have no problem in following the expression expansion rule and thus pick style \#1.

We also have the decision of whether or not we want to put a comma after the last argument. But not having a trailing comma on the last argument would be even worse than the leading comma style, remember that's how we got into this situation in the first place! (The leading comma style was a lesser-evil workaround back in the days where the language didn't allow a trailing comma on the last item.)

### 2. Multiline arguments

There is little to debate here. The syntax looks so similar to bindings in attrsets, that we can simply copy the rules over 1:1 (replacing `=` with `?` and `;` with `,`). Yay for consistency!

```nix
# 1
{
  one ? [
    1
    1
  ],
  two ?
    if "two" == 2 then 2 else "two"
  three ? function [
    1
    2
  ],
}:

[]
```


### 3. `@`-pattern

Our first non-trivial decision with trailing commas.

```nix
# 1
args@{
  one,
  two,
  three,
}:

{
  one,
  two,
  three,
}@args:

# 2
args @ {
  one,
  two,
  three,
}:

{
  one,
  two,
  three,
} @ args:

# 3
args@
{
  one,
  two,
  three,
}:

{
  one,
  two,
  three,
}
@args:

# 4
args @
{
  one,
  two,
  three,
}:

{
  one,
  two,
  three,
}
@ args:

[]
```

Leading `@`-binding looks better than the trailing one, but it is more common in Nixpkgs anyways. We could try to normalize to always use leading `@`-bindings, but if possible this should be avoided so let's discard that option.

For consistency, the leading and trailing version should look the same. There are many possibilities on how to place the whitespace here. I'll optimize for the more common leading `@` style and pick option \#1 or \#2.

### 4. Comments

Nothing really to discuss here either. Again, we'll just copy the style over 1:1 from bindings.

```nix
# 1
{
  /* comment */
  one,
  /*
    multiline
    comment
  */
  two,
  # comment
  three,
  # multiline
  # comment
  four,
}:

[]
```


## Conclusion

Yes, we can somehow get leading commas to work. But at every corner we are faced with tough decisions between sub-optimal options, having to make exceptions to rules and allowing inconsistencies. The only argument in favor of leading commas would be the widespread usage in current Nixpkgs. However, the RFC states on this topic:

> When deciding between two *equally good* options, currently prevalent formatting style in Nixpkgs should be followed. The emphasis here is on "equally good". We should not fear making radical changes to the current style if there are sufficient arguments in favor of it.

As we've just seen, both options are not "equally good". To me, trailing commas are good but leading commas are *okay at best*.

And that's the reason why the trailing commas style was chosen in this RFC.